include/boost/geometry/algorithms/detail/partition.hpp

```diff
@@ -589,53 +653,57 @@ public :
 
     template
     <
-        typename InputCollection1,
-        typename InputCollection2,
+        typename ForwardRange1,
+        typename ForwardRange2,
         typename VisitPolicy
     >
-    static inline void apply(InputCollection1 const& collection1,
-                InputCollection2 const& collection2,
+    static inline void apply(ForwardRange1 const& forward_range1,
+                ForwardRange2 const& forward_range2,
                 VisitPolicy& visitor,
                 std::size_t min_elements = 16,
-                VisitBoxPolicy box_visitor = detail::partition::visit_no_policy()
+                VisitBoxPolicy box_visitor
+                    = detail::partition::visit_no_policy()
                 )
     {
-        if (std::size_t(boost::size(collection1)) > min_elements
-            && std::size_t(boost::size(collection2)) > min_elements)
+        typedef typename boost::range_iterator
+            <
+                ForwardRange1 const
+            >::type iterator_type1;
+
+        typedef typename boost::range_iterator
+            <
+                ForwardRange2 const
+            >::type iterator_type2;
+
+        if (std::size_t(boost::size(forward_range1)) > min_elements
+            && std::size_t(boost::size(forward_range2)) > min_elements)
         {
-            index_vector_type index_vector1, index_vector2;
+            std::vector<iterator_type1> iterator_vector1;
+            std::vector<iterator_type2> iterator_vector2;
```

Comment:
Regarding the size: it depends on the size of the iterator of course.

My understanding is that for the existing use cases we should not see a difference in the size of the vector: what was passed was a random access container/range and an integer index; now what I pass is an iterator to the random access container/range, and I expect the size of this iterator to be the same as the size of the index.

It is true that if you pass a possibly more complicated iterator (like `point_iterator<>` or `segment_iterator<>`) then the size of the vector may be larger; on the other hand such usages where not possible before, and in that respect I do not see an issue with respect to size.

I will investigate this in more detail as part of the performance comparison.